# ==============================================================================
# FIXING COMMIT DIFF
# ==============================================================================
# File: security/sandbox/linux/SandboxFilter.cpp
# Commit: 9c0f809e9924
# Full Hash: 9c0f809e99246bcd7636a117ae50d67ab719471d
# Author: Jed Davis <jld@mozilla.com>
# Date: 2020-07-14 03:18:58
# Description:
#   Bug 1650751 - Add FMODE_NONOTIFY to ignored file flags in Linux sandbox. r=gcp
#   
#   As of kernel 5.8 (commit [e9c15badb][]), Linux will set the internal
#   `FMODE_NONOTIFY` flag on files that don't exist in the filesystem,
#   including (unnamed) pipes and sockets.  Although this flag isn't
# ==============================================================================

diff -r 918de9580b2a -r 9c0f809e9924 security/sandbox/linux/SandboxFilter.cpp
--- a/security/sandbox/linux/SandboxFilter.cpp	Mon Jul 13 21:35:22 2020 +0000
+++ b/security/sandbox/linux/SandboxFilter.cpp	Mon Jul 13 12:18:00 2020 +0000
@@ -70,6 +70,9 @@
 // actual value because it shows up in file flags.
 #define O_LARGEFILE_REAL 00100000
 
+// Not part of UAPI, but userspace sees it in F_GETFL; see bug 1650751.
+#define FMODE_NONOTIFY 0x4000000
+
 #ifndef F_LINUX_SPECIFIC_BASE
 #  define F_LINUX_SPECIFIC_BASE 1024
 #else
@@ -541,7 +544,7 @@
         // new SETFL-able flags are added.  (In particular we want to
         // forbid O_ASYNC; see bug 1328896, but also see bug 1408438.)
         static const int ignored_flags =
-            O_ACCMODE | O_LARGEFILE_REAL | O_CLOEXEC;
+            O_ACCMODE | O_LARGEFILE_REAL | O_CLOEXEC | FMODE_NONOTIFY;
         static const int allowed_flags = ignored_flags | O_APPEND | O_NONBLOCK;
         return Switch(cmd)
             // Close-on-exec is meaningless when execve isn't allowed, but
